### PR TITLE
Stochastic VMP with natural gradient descent

### PR DESCRIPTION
**Pull request by semihakbayrak**
- This PR implements two well recognised natural gradient descent variational inference methods, Stochastic Variational Inference (SVI by Hoffman et al) and Conjugate-computation Variational Inference (CVI by Khan and Lin), for probabilistic programming on factor graphs as message passing algorithms. 
- This is achieved through SVI and CVI nodes. SVI node aims to scale variational inference in conjugate models to large datasets, which obviates the need for generating the entire graph and therefore speeds up the algorithm generation procedure of ForneyLab for i.i.d. datasets. With the help of SVI node, variational inference can be executed in a fully online manner or with mini-batches. 
- CVI node generalizes the natural gradient descent approach to non-conjugate model specifications and extends the universality of ForneyLab. 
- Both SVI and CVI methods are developed having exponential family of distributions in mind. Therefore, this PR also enriches probability distributions in ForneyLab with their canonical representations. 
- For CVI node, estimation of gradients and Fisher Information Matrices (FIM) are handled by ForwardDiff and Zygote autodiff packages of Julia language. In the current implementation, we use score function estimators for gradient estimation if the approximate posterior family is not Gaussian. For SVI, natural gradient steps nicely interface with messages and are available in closed form.
- Optimizers to be used in stochastic approximations can be chosen among Flux.Optimise options. An additional ForgetDelayDescent optimizer that is proposed in SVI paper is implemented.

### CODE REVIEW

**Review by albertpod — COMMENTED** [review]
Thanks for PR @semihakbayrak! It's a very big contribution to `ForneyLab`.
You may have some demos you'd want to add to this PR?

File: src/ForneyLab.jl
Comment (albertpod) [inline src/ForneyLab.jl:16]: Any chance we can avoid `using Zygote`?
It will affect the compilation speed of ForneyLab.

File: src/engines/julia/update_rules/cvi.jl
Comment (albertpod) [inline src/engines/julia/update_rules/cvi.jl:7]: This is dangerous. What if active graph changes in a global environment?
But I understand you can't pass a node object to this function. Maybe @ThijsvdLaar can see a better solution here.

File: src/engines/julia/update_rules/cvi.jl
Comment (albertpod) [inline src/engines/julia/update_rules/cvi.jl:132]: You overwrite the method here. This is undesirable. Julia throws a warning, i.e.
```julia
WARNING: Method definition (::ForneyLab.var"#logp_nc#364"{msg_out, arg_sample, inx_list, thenode, j})(Any) in module ForneyLab at /Users/apodusenko/.julia/dev/ForneyLab/src/engines/julia/update_rules/cvi.jl:129 overwritten at /Users/apodusenko/.julia/dev/ForneyLab/src/engines/julia/update_rules/cvi.jl:132.
  ** incremental compilation may be fatally broken for this module **
```

One way to circumvent this warning was adviced by @bvdmitri:
```julia
logp_nc = (z) -> sum(logPdf.([msg_out.dist],thenode.g.(arg_sample(z,inx_list[j])...)))/thenode.num_samples
if thenode.online_inference[inx_list[j]]
    λ_init = deepcopy(naturalParams(thenode.q[inx_list[j]]))
    logp_nc = (z) -> (thenode.dataset_size/thenode.batch_size)*sum(logPdf.([msg_out.dist],thenode.g.(arg_sample(z,inx_list[j])...)))/thenode.num_samples
else
    λ_init = deepcopy(naturalParams(msg_in.dist))
end
```
If we enter the `if statement,` two functions will be created. The first one will be handled by a garbage collector.

Comment (bvdmitri) [reply src/engines/julia/update_rules/cvi.jl:132]: Another option would be:

```julia
(logp_nc, λ_init) = if thenode.online_inference[inx_list[j]]
    λ_init = deepcopy(naturalParams(thenode.q[inx_list[j]]))
    logp_nc = (z) -> (thenode.dataset_size/thenode.batch_size)*sum(logPdf.([msg_out.dist],thenode.g.(arg_sample(z,inx_list[j])...)))/thenode.num_samples
    (logp_nc, λ_init)
else
    λ_init = deepcopy(naturalParams(msg_in.dist))
    logp_nc = (z) -> sum(logPdf.([msg_out.dist],thenode.g.(arg_sample(z,inx_list[j])...)))/thenode.num_samples
if thenode.online_inference[inx_list[j]]
    (logp_nc, λ_init)
end
```

`if` statement in Julia returns an actual last value from executed branch, so you can use it as ternary operator:

```julia
value = if some_condition
    0.0
else
    1.0
end
```

File: src/ForneyLab.jl
Comment (bvdmitri) [reply src/ForneyLab.jl:16]: @albertpod @semihakbayrak https://github.com/JuliaPackaging/Requires.jl

Comment (bvdmitri) [reply src/ForneyLab.jl:16]: This package allows you to extend package's functionality if other packages (e.g. `Zygote`) are loaded already and at the same time the original package does not depend on `Zygote`.

**Comment by semihakbayrak** [comment]
> Thanks for PR @semihakbayrak! It's a very big contribution to `ForneyLab`.
> You may have some demos you'd want to add to this PR?

Thanks @albertpod! I'll add some example notebooks on Friday.

File: src/engines/julia/update_rules/cvi.jl
Comment (semihakbayrak) [reply src/engines/julia/update_rules/cvi.jl:132]: I couldn't make it work without overwriting the method at first place. But I'll consider @bvdmitri's suggestions.

**Review by ThijsvdLaar — CHANGES_REQUESTED** [review]
Thanks for all the work @semihakbayrak . I think conceptually the SVI and CVI nodes present an interesting and important contribution. Unfortunately, as a ForneyLab maintainer, I see four major issues that need to be addressed before this work can be merged into the ForneyLab core.

1) The PR consistes of multiple contributions: CVI and SVI nodes (with auxiliary definitions), and the introduction of natural parameterizations. These are separate contributions and as such should each have their own PR. This will simplify discussions and feedback, and also ensure that implementations themselves are as modular as possible.

2) The CVI and SVI nodes implicitly introduce a plate notation in order to work with batches. The way in which this is currently implemented conflates model definition with the inference definition, and it breaks locality for model definition and inference. Furthermore, in order to handle the complications introduced by these nodes, several auxiliary concepts are introduced, i.e. `set_prob_dist`, `set_sample_list` and `joint_indep_prob_dist`. To me, this indicates deeper conceptual problems with the plate concept in the context of FFGs that go beyond the contributions presented in this PR. The PR would thus intruduce a significant technical debt, which cannot swiftly be resolved. Before these contributions can be merged in the FL core, model definition and inference for these nodes should be separated out. For this, we might consider leveraging the concept of a composite node to move the batch updating procedures as a part of a more efficient update.

3) Naming and consistency should be improved throughout the code base. The CVI node is in essence a `NonLinear` node that computes messages by gradient descent on the natural parameters. In essence, this makes is an additional approximation method, `Nonlinear{CVI}`, alongside `Nonlinear{Sampling}` etc. Here, the `NonLinear` node naming might not be appropriate anymore, because it also applies to linear transformations. Therefore this contribution might also affect other naming conventions, which should be considered.

4) Several updates refer to the `currentGraph()`. This prohibits standalone execution of the algorithm code.

**Comment by ThijsvdLaar** [comment]
Changes are implemented by #192, closing this PR.

**Pull request closed without merging**
